src/jetbrains/buildServer/tools/FilesProcessor.java

```diff
@@ -22,59 +22,131 @@
-      private final ScanStep[] steps = new ScanStep[]{
-              new CheckFileScanStep(settings, reporting),
-              new ScanZipStep(),
-              new DirectoryScanStep(),
-      };
+    try {
+      new FilesProcessor(settings, reporting).processRoot(scanHome);
+    } catch (Exception e) {
+      if (e instanceof RuntimeException) throw ((RuntimeException) e);
+      throw new RuntimeException(e);
+    }
+  }
 
-      private int cnt = 0;
 
-      public void postTask(@NotNull ScanFile file) {
-        if (settings.isDebugMode()) {
-          System.out.println("scanning: " + file.getName());
-        } else {
-          cnt++;
-          if (cnt % 100 == 0) {
-            System.out.print(".");
-          }
-          if (cnt % 4000 == 0) {
-            System.out.println();
-          }
+  private final ThreadPoolExecutor myExecutor = new ThreadPoolExecutor(1, Runtime.getRuntime().availableProcessors(), 1, TimeUnit.MINUTES, new LinkedBlockingDeque<Runnable>()) {
+    {
+      setThreadFactory(new ThreadFactory() {
+        @Override
+        public Thread newThread(@NotNull Runnable r) {
+          final Thread thread = new Thread(r);
+          thread.setName("Files Processor " + ourThreadNameCounter.getAndIncrement());
+          return thread;
         }
+      });
+    }
+  };
+  private final Queue<Future> myFutures = new ConcurrentLinkedQueue<Future>();
 
-        if (settings.isPathExcluded(file)) {
-          System.out.print("S");
-          return;
-        }
+  private final AtomicInteger myProcessed = new AtomicInteger();
+  @NotNull
+  private final CheckSettings mySettings;
+  @NotNull
+  private final ErrorReporting myReporting;
+  private final ScanStep[] mySteps;
 
-        for (ScanStep step : steps) {
-          try {
-            step.process(file, this);
-          } catch (Throwable e) {
-            reporting.postError(file, e.toString());
-          }
-        }
-      }
+  private FilesProcessor(@NotNull final CheckSettings settings,
+                         @NotNull final ErrorReporting reporting) {
+    mySettings = settings;
+    myReporting = reporting;
+    mySteps = new ScanStep[]{
+            new CheckFileScanStep(mySettings, myReporting),
+            new ScanZipStep(),
+            new DirectoryScanStep(),
     };
+  }
 
-    c.postTask(new FSScanFileBase(scanHome){
+  private void processRoot(@NotNull final File scanHome) throws ExecutionException, InterruptedException {
+    final FSScanFileBase base = new FSScanFileBase(scanHome) {
       @NotNull
       @Override
       public String getName() {
         return "";
       }
-    });
+    };
+    this.postTask(base);
+
+    if (!myFutures.isEmpty()) {
+      myExecutor.prestartAllCoreThreads();
+    }
+    // Wait till all files processed
+    while (!myFutures.isEmpty()) {
+      final Iterator<Future> it = myFutures.iterator();
+      Future lastNonDone = null;
+      while (it.hasNext()) {

```

Comment:
ConcurrentModificationException?

Reply:
Sure, just copied it from non-concurrent similar code in IntelliiJ build scripts